src/logging.c

```diff
@@ -1,49 +1,91 @@
 #include <stdarg.h>
 #include <time.h>
+#include <string.h>
+#include <libgen.h>
+#include <stdlib.h>
+#include <SDL2/SDL.h>
 
 #include "logging.h"
+#include "rom.h"
 
 /******************************************************
  *** LOCAL VARIABLES                                ***
  ******************************************************/
 
+#define MAX_TIME_STR_LEN (20)
+
 static LoggingLevel min_log_lvl = FATAL;
 
+/******************************************************
+ *** LOCAL METHODS                                  ***
+ ******************************************************/
+
+static void get_time(char* const time_str) {
+	time_t current_time;
+	struct tm local_time;
+	time(&current_time);
+	localtime_r(&current_time, &local_time);
+	strftime(time_str, MAX_TIME_STR_LEN, "%F %T", &local_time);
+}
+
 /******************************************************
  *** EXPOSED METHODS                                ***
  ******************************************************/
 
 void log_set_lvl(LoggingLevel log_lvl) {
     min_log_lvl = log_lvl;
 
-    // minimum debug level cannot be higher than fatal
-    if (min_log_lvl > FATAL)
-        min_log_lvl = FATAL;
+    // minimum logging level cannot be higher than fatal
+	if (min_log_lvl > FATAL) {
+		LOG_WARNING("Provided logging level %d is higher than FATAL (%d).", min_log_lvl, FATAL);
+		min_log_lvl = FATAL;
+	}
 
-    // minimum debug level cannot be lower than debug
-    if (min_log_lvl < DEBUG)
-        min_log_lvl = DEBUG;
+    // minimum logging level cannot be lower than debug
+	if (min_log_lvl < DEBUG) {
+		LOG_WARNING("Provided logging level %d is lower than DEBUG (%d).", min_log_lvl, DEBUG);
+		min_log_lvl = DEBUG;
+	}
 
     LOG_INFO("Log level initialized to %d.", min_log_lvl);
 }
 
-__attribute__((format(printf, 4, 5)))
-void log_str(LoggingLevel const log_lvl, const char *const log_lvl_str, FILE *const stream, const char *const msg, ...) {
-    if (log_lvl < min_log_lvl) return;
+void log_exit(const char *const file_path, const int line_number, const char *const msg, ...) {
+	va_list args;
+	va_start(args, msg);
 
-    va_list args;
-    time_t current_time;
-    struct tm local_time;
-    char time_str[20];
+	char time_str[MAX_TIME_STR_LEN];
+	get_time(time_str);
 
-    va_start(args, msg);
+	char *file_path_copy = strdup(file_path);
+	const char *file_name = basename(file_path_copy);
 
-    time(&current_time);
-    localtime_r(&current_time, &local_time);
-    strftime(time_str, 20, "%F %T", &local_time);
+	fprintf(stderr, "%10s – %s – %s:%d - ", "FATAL", time_str, file_name, line_number);
+	vfprintf(stderr, msg, args);
+	fprintf(stderr, "\n");
```

Comment:
I thought about that too. There are some instances where we have `LOG_EXIT("something %s because %d", my_str, x);`. That's why we cannot avoid variadic arguments (as far as I can see). But perhaps I'm overseeing something here 🤔

Reply:
One solution to this could be to use snprintf in `log_exit`.

Reply:
How about replacing the `#define LOG_EXIT` with
```C
#define LOG_EXIT(msg, ...) LOG_FATAL(msg, ##__VA_ARGS__); exit(1)
``` 
in combination with the supposed usage of https://linux.die.net/man/3/atexit.

Reply:
We'll leave this for now.

As a side note, one possible solution could be to make `LOG_*` print file name and line number. But this is not desired atm.